tests/units/plugins/filter/test_merge.py

```diff
@@ -0,0 +1,258 @@
+"""
+Tests for the "merge" plugin filter
+"""
+from copy import deepcopy
+
+from ansible.compat.tests import unittest
+from ansible.errors import AnsibleFilterError
+
+from plugins.filter.merge import merge_with_app
+
+
+def deep_sort_dict(d):
+    """
+        Recursively alphabetically sort lists as dictionary values (input dict
+        is not modified)
+    """
+    for k, v in d.iteritems():
+        if isinstance(v, list):
+            for i in v:
+                if isinstance(i, dict):
+                    deep_sort_dict(i)
+            d[k] = sorted(v)
+        if isinstance(v, dict):
+            deep_sort_dict(v)
+    return d
+
+
+class TestDeepSortDict(unittest.TestCase):
+    """Test the ``deep_sort_dict`` utility"""
+
+    def test_flat_dict(self):
+        """Avoid testing recusive calls"""
+
+        d = {"foo": 1, "bar": ["c", "b", "d", "a"]}
+        expected = {"foo": 1, "bar": ["a", "b", "c", "d"]}
+
+        assert deep_sort_dict(d) == expected
+
+    def test_embedded_dict(self):
+        """Test recursive calls"""
+
+        d = {"foo": {"baz": [4, 5, 6, 1]}, "bar": ["c", "b", "d", "a"]}
+        expected = {"foo": {"baz": [1, 4, 5, 6]}, "bar": ["a", "b", "c", "d"]}
+
+        assert deep_sort_dict(d) == expected
+
+    def test_deep_embedded_dict(self):
+        """Avoid testing recursivity"""
+
+        d = {
+            "foo": {"baz": [4, 1, 6, {"lol": ["x", "z", "y"]}]},
+            "bar": ["c", "b", "d", "a"],
+        }
+        expected = {
+            "foo": {"baz": [1, 4, 6, {"lol": ["x", "y", "z"]}]},
+            "bar": ["a", "b", "c", "d"],
+        }
+
+        assert deep_sort_dict(d) == expected
```

Comment:
you should add:

assert d == {
    "foo": {"baz": [4, 1, 6, {"lol": ["x", "z", "y"]}]},
    "bar": ["c", "b", "d", "a"],
}

Same on all tests?

Reply:
Not if we test immutability separately?